### ISSUE TITLE
[45] Masked Feature Prediction for Self-Supervised Visual Pre-Training (MaskFeat)

### ISSUE THREAD

**Issue by dhkim0225** [dhkim0225/1day_1paper#73]
microsoft의 [SimMIM](https://arxiv.org/abs/2111.09886) 과 굉장히 비슷함.
decoder 하나 두고,
masked patch 함께 encoder 에 넣어주고,
다른 점은, HOG feature 를 예측한다는 것 !

target task ==> video benchmarks
pretrain 때도 video 데이터 어떻게 쓸 지 고려함.

[paper](https://arxiv.org/pdf/2112.09133.pdf)
[code](https://github.com/facebookresearch/pytorchvideo) ==> 곧 업로드 한다고 적혀 있음

3가지 결론
(i) HOG 나 SIFT 예측하게 하는게 꽤 효과적
(ii) visual prediction을 위해 굳이 discretize 할 필요 없음
(iii) local pattern 을 characterize 하는 게 중요함. pretrained CNN이나 pretrained ViT 의 feature 는 학습시키면 성능 더 구려짐.

# Method
![image](https://user-images.githubusercontent.com/16400591/146865025-8a500600-271c-4180-8be3-2c70a183a207.png)
아 ㅋㅋ 그림 하나면 끝난다구~

## Target Features
1. RGB pixle (l2 loss)
2. HOG feature (l2 loss)
3. 8192 dVAE token (CE loss)
4. Deep features (cosine_distance. i.e. MSE of l2-normalized features)
    1. cnn, vit 등 ImageNet pretrained 의 feature 예측하게 함.
5. Pseudo Label
    1. [Token-Labelling](https://arxiv.org/abs/2104.10858) 사용
    2. label 생성 네트워크는 NF-Net 사용

# Results
### Kinetics-400 (K400)
![image](https://user-images.githubusercontent.com/16400591/146865348-8d3aca32-5ca2-4f71-8f7d-e94f4b7ce09f.png)
ViT-B, IN-1K 에서 pretraining => Kinetics-400 에서 finetune

### ImageNet 1K
![image](https://user-images.githubusercontent.com/16400591/146865404-edf9bc75-7b13-4164-93fd-c3dc2351450a.png)
논문들 수치 가져온 게 아니라, MaskFeat 에 저 feature 들 다 넣어서 pretraining 해보고 결과 report 한 것임.
ViT-B, IN-1K 에서 pretraining => IN-1K 에서 finetune

~SimMIM 에서 RGB score 는 좀 더 높긴 함. Masking ratio에서 차이남. 여기서는 40% mask.~

## Kinetics-400
![image](https://user-images.githubusercontent.com/16400591/146873293-62fc29f4-68ea-4873-b6db-f303716e1cbe.png)
재밌는 점은 IN-21K pretrain 도 Kinetic 에서는 힘을 못 쓴다는 점.

### masking
- frame
    - 프레임 별로 다른 masking 전략 취하기
- tube
    - 2d mask map 을 뽑고 temporal 방향으로 (시간축으로) 쭉 같이 사용함
- gray
    -  random time step 까지만 tube 를 적용함.
    - 다른 time step 부터 다시 위 작업을 loop 돌면서 수행함.
![image](https://user-images.githubusercontent.com/16400591/146874180-50c78579-92f4-41ec-8113-b38b56a99169.png)

### schedule
큰 모델일수록 오래 학습시키는게 효과적
![image](https://user-images.githubusercontent.com/16400591/146874623-e265fed6-1c82-4d30-b42d-4e374de90319.png)

### Target design
video 데이터 pretrain 시에 mask 부분의 hog 에 loss 를 주냐,
전체 cube 에 대해 loss 를 주냐 ablation 해보니까,
각각 주는게 낫더라.
![image](https://user-images.githubusercontent.com/16400591/146873914-d5a2e53b-ffd9-44c1-83c8-211e32c49241.png)
